### ISSUE TITLE
DUPLICATE_NAME_ERROR when training GAN

### ISSUE THREAD

**Issue by hughplay** [uber/horovod#784]
There is no problem with this great framework while training a normal model.

But yesterday when I attempt to train a GAN, Horovod raise RuntimeError. 

I guess two optimizers are conflicted.

```
RuntimeError: Horovod has been shut down. This was caused by an exception on one of the ranks or an attempt to allreduce, allgather or broadcast a tensor after one of the ranks finished execution. If the shutdown was caused by an exceptio
n, you should see the exception in the log before the first shutdown message.
```

And I found this error is defined as `DUPLICATE_NAME_ERROR ` in [horovod/horovod/common/operations.cc](https://github.com/uber/horovod/blob/c314949a60fca30674388b69e730c83b79d22ff5/horovod/common/operations.cc)

Optimizer definition:

``` py
self.generator_optimizer = get_optimizer(self.config, gen_parameters, hvd)
self.discriminator_optimizer = get_optimizer(self.config, disc_parameters, hvd)
self.generator_optimizer = hvd.DistributedOptimizer(self.generator_optimizer, named_parameters=gen_parameters)
self.discriminator_optimizer = hvd.DistributedOptimizer(self.discriminator_optimizer, named_parameters=disc_parameters)
```

Errors raised when `g_loss.backward()`:

``` py
self.generator.train()
self.discriminator.train()

gen_imgs = self.generator(sketches)

self.generator_optimizer.zero_grad()
g_loss = self.criterion(as_real=self.discriminator(gen_imgs))
g_loss.backward()
self.generator_optimizer.step()

self.discriminator_optimizer.zero_grad()
d_loss = self.criterion(as_real=self.discriminator(imgs), as_fake=self.discriminator(gen_imgs.detach()))
d_loss.backward()
self.discriminator_optimizer.step()
```

Could you please tell me how can I solve this problem?

**Comment by alsrgv** [comment]
@hughplay, I suspect that the issue is here:
```python
gen_imgs = self.generator(sketches)

self.generator_optimizer.zero_grad()
g_loss = self.criterion(as_real=self.discriminator(gen_imgs))
g_loss.backward()
self.generator_optimizer.step()
```

Do you intend this step to train both generator and discriminator?  If not, you need to detach generator outputs, or [flip `requires_grad_`](https://github.com/uber/horovod/blob/master/test/test_torch.py#L1182).

If you do intend to train both of them concurrently, you'll need to make sure that variable names of generator and discriminator do not overlap.  E.g. if both of them has a layer named `conv`, this will be a duplicate that causes this issue.  You do could that by introducing a proxy nn.Model that would disambiguate the names, or by hacking `named_parameters()`.

**Comment by hughplay** [comment]
@alsrgv Thank you for you reply.

The code you highlight is to optimize generator. 

When optimizing discriminator, I `detach` the `gen_imgs`.

``` py
self.discriminator_optimizer.zero_grad()
d_loss = self.criterion(as_real=self.discriminator(imgs), as_fake=self.discriminator(gen_imgs.detach()))
d_loss.backward()
self.discriminator_optimizer.step()
```

In fact, the reference code is here: https://github.com/eriklindernoren/PyTorch-GAN/blob/master/implementations/dcgan/dcgan.py

So should I "freeze" discriminator's parameters when backward g_loss? And "unfreeze" it after  `self.generator_optimizer.step()`?

I will try that and may be do some hacking as you said above. Thank you again.

**Comment by hughplay** [comment]
I have resolved this problem after carefully reading the examples and document in source code. :stuck_out_tongue_closed_eyes:

The DUPLICATE_NAME_ERROR raised because I didn't give a correct `named_parameters` to `hvd. DistributedOptimizer `. I passed `model.parameters()` rather than `model.named_parameters()`.

 So when process the line below, both `generator_optimizer` and `discriminator_optimizer` will have this default format to construct their parameter names for the naming of all_reduce operation. Thus lead to "duplicate".
https://github.com/uber/horovod/blob/42c0977d8ee62060b1d4c25a98b15c9cac85ebef/horovod/torch/__init__.py#L62

However, there is another problem which makes me confused. Because that problem is about the backward hook, I will close this issue and open a new issue.

Thank you.
